src/test/crimson/seastore/transaction_manager_test_state.h

```diff
@@ -111,21 +73,56 @@ class EphemeralTestState {
             segment_manager::get_ephemeral_device_config(cnt, get_num_devices()));
         });
       });
-    }).safe_then([this] {
-      _init();
-      return _mkfs();
-    }).safe_then([this] {
-      return restart_fut();
     }).handle_error(
       crimson::ct_error::assert_all{}
     ).then([FNAME] {
       SUBINFO(test, "finish");
     });
   }
 
-  seastar::future<> randomblock_setup()
-  {
-    LOG_PREFIX(EphemeralTestState::randomblock_setup);
+  void remount() {
+    segment_manager->remount();
+    for (auto &sec_sm : secondary_segment_managers) {
+      sec_sm->remount();
+    }
+  }
+
+  std::size_t get_num_devices() const {
+    return secondary_segment_managers.size() + 1;
+  }
+
+  void reset() {
+    segment_manager.reset();
+    for (auto &sec_sm : secondary_segment_managers) {
+      sec_sm.reset();
+    }
+  }
+
+  std::vector<Device*> get_secondary_devices() {
+    std::vector<Device*> sec_devices;
+    for (auto &sec_sm : secondary_segment_managers) {
+      sec_devices.emplace_back(sec_sm.get());
+    }
+    return sec_devices;
+  }
+
+  segment_manager::EphemeralSegmentManager* get_primary_device() {
+    return segment_manager.get();
+  }
+};
+
+class EphemeralRandomBlockDevices : public EphemeralDevices {
+  random_block_device::RBMDeviceRef rb_device;
+  std::list<random_block_device::RBMDeviceRef> secondary_rb_devices;
+
+public:
+  EphemeralRandomBlockDevices(std::size_t num_device_managers) {
+    assert(num_device_managers > 0);
+    secondary_rb_devices.resize(num_device_managers);
```

Comment:
Although num_device_manager is one, the RBM test needs two random block devices for now---one is for a primary device (CBJournal), and another is for a secondary device (RBM).

Reply:
hm.. Although a single device is used, don't we normally use the device with partitioning? For instance, sda device --> sda1, sda2. So, I thought that the reasonable way is that CBJournal uses the sda1 and RBM uses sda2. Do you mean a device represents a entire device, not a partition? If so, we need a partition table to find where the RBMs begin, which is the same way as existing partition method I mentioned.

Reply:
> Do you mean a device represents a entire device, not a partition?

Yes, I think a device represents a raw device conceptually.

> If so, we need a partition table to find where the RBMs begin, which is the same way as existing partition method I mentioned.

We already use software-defined superblock to remember the disk ranges for different purposes, for example https://github.com/ceph/ceph/blob/main/src/crimson/os/seastore/segment_manager.h#L30, and I think its simpler to configure the RBM journal size rather than requiring a separate step to do disk partitioning, which also needs to remember the external partitions.

What do you think?

Reply:
It is ok as is. But, we probably will need a kind of table in superblock if a device includes several RBMs. Also, need a protection mechanism to prevent accessing other device regions.

Reply:
I'm wondering if considering multiple RB devices is adding unnecessary complexities in the current step.

IIUC if SeaStore supports only one RB device at the time, users can still combine multiple RB devices as one using other methods, right?

If that's true, probably its simpler to focus on enabling RBM with single device as the first step (e.g. drop `RBMDeviceGroup` as well).

>  But, we probably will need a kind of table in superblock if a device includes several RBMs.  Also, need a protection mechanism to prevent accessing other device regions.

I don't follow why a device will need several RBMs? I might be missing something, but seems currently the superblock only need to distinguish two static ranges -- circular journal and ool extents.

Reply:
> IIUC if SeaStore supports only one RB device at the time, users can still combine multiple RB devices as one using other methods, right?

Right.

> I don't follow why a device will need several RBMs? I might be missing something, but seems currently the superblock only need to distinguish two static ranges -- circular journal and ool extents.

There are some uses cases in classic OSD scenario where an user make several partitions, each of which has an object store (bluestore) because a single objecstore (with a NVMe device) and an OSD fails to reach NVMe's performance. Therefore, I wonder if we need to consider this case as well. If not, we can move forward without considering this.

Reply:
> There are some uses cases in classic OSD scenario where an user makes several partitions on a NVMe device, each of which has an object store (bluestore) because a single objecstore (with a NVMe device) and an OSD fails to reach NVMe's performance. Therefore, I wonder if we need to consider this case as well. If not, we can move forward without considering this.

Crimson/SeaStore is trying to avoid that use case with the shared-nothing design to scale OSD performance with CPUs. That should allow a single Crimson OSD to consume the entire NVMe device without the partitioning trick.